# Código y trabajo

Parte 3
-----

a)
--

In [23]:
import pandas as pd
from sklearn.cross_validation import train_test_split

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data'
df = pd.read_csv(url, sep='\s+', header=None, names=['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV'])
df_train, df_test = train_test_split(df, test_size=0.2, random_state=0)
df.shape
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null int64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null int64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
MEDV       506 non-null float64
dtypes: float64(12), int64(2)
memory usage: 55.4 KB


CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677082   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

El dataset a utilizar es un archivo con 506 distitos datos, separados con 14 distitas características. Se puede observar en la tabla mostrada anteriormente un resumen de los distintos valores estadísticos de cada una de las columnas para todos los valores.

b)
--


In [24]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(df_train)
X_train_scaled = pd.DataFrame(scaler.transform(df_train), columns=df_train.columns)
y_train_scaled = df_train.pop('MEDV')

In [25]:
X_test_scaled = pd.DataFrame(scaler.transform(df_test), columns=df_test.columns)
y_test_scaled = df_test.pop('MEDV')

c)
--

In [26]:
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD

model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

sgd = SGD(lr=0.01)
model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))


/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:6: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:8: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: 94.7232 - val_loss: 46.9930
Epoch 2/300
404/404 [==============================] - 0s - loss: 30.0625 - val_loss: 35.6402
Epoch 3/300
404/404 [==============================] - 0s - loss: 20.7890 - val_loss: 29.2223
Epoch 4/300
404/404 [==============================] - 0s - loss: 17.7693 - val_loss: 24.5414
Epoch 5/300
404/404 [==============================] - 0s - loss: 13.7015 - val_loss: 20.3739
Epoch 6/300
404/404 [==============================] - 0s - loss: 11.4330 - val_loss: 17.6146
Epoch 7/300
404/404 [==============================] - 0s - loss: 9.9379 - val_loss: 1

Claramente se puede observar que hay una convergencia, esto se ve en el nivel de pérdida, partiendo con un valor superior a 99 llegando a un nivel de pérdida de 0.0840, lo cual demuestra claramente la convergencia del entrenamiento.

d)
--

In [27]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

sgd = SGD(lr=0.01)
model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))



/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: 235.2082 - val_loss: 32.3378
Epoch 2/300
404/404 [==============================] - 0s - loss: 14.4019 - val_loss: 13.4594
Epoch 3/300
404/404 [==============================] - 0s - loss: 5.9000 - val_loss: 7.5723
Epoch 4/300
404/404 [==============================] - 0s - loss: 3.4907 - val_loss: 5.2605
Epoch 5/300
404/404 [==============================] - 0s - loss: 2.5463 - val_loss: 5.0126
Epoch 6/300
404/404 [==============================] - 0s - loss: 1.8561 - val_loss: 2.5660
Epoch 7/300
404/404 [==============================] - 0s - loss: 1.4801 - val_loss: 2.255

Para este caso también se observa una convergencia, sin embargo es mucho más inestable, aunque igual converge mucho más rápido que en el punto c.

e)
--

In [28]:
import numpy as np

n_lr = 20
learn_rate = np.linspace(0,1,n_lr)

In [12]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

sgd = SGD(lr=learn_rate)
model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))



/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


ValueError: Dimension 0 in both shapes must be equal, but are 1 and 200 for 'Assign_21' (op: 'Assign') with input shapes: [1], [200].

In [14]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

sgd = SGD(lr=learn_rate)
model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))




/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


ValueError: Dimensions must be equal, but are 20 and 200 for 'mul_39' (op: 'Mul') with input shapes: [20], [200].

f)
--

In [9]:
from sklearn import cross_validation
import numpy as np
K=5
Xm = X_train_scaled.as_matrix()
ym = y_train_scaled.as_matrix()
kfold = cross_validation.KFold(len(Xm),K)
cvscores = []

for i, (train,val) in enumerate(kfold):
    model = Sequential()
    model.add(Dense(200, input_dim=Xm.shape[5], init='uniform'))
    model.add(Activation('relu'))
    model.add(Dense(150, init='uniform'))
    model.add(Activation('linear'))
    model.add(Dense(100, init='uniform'))
    model.add(Activation('linear'))
    model.add(Dense(150, init='uniform'))
    model.add(Activation('linear'))
    model.add(Dense(1, init='uniform'))
    model.add(Activation('linear'))
    
    sgd = SGD(lr=0.2)
    model.compile(optimizer=sgd, loss='mean_squared_error')
    
    model.fit(Xm[train], ym[train], nb_epoch=300)
    
    scores = model.evaluate(Xm[val], ym[val])
    cvscores.append(scores)
mse_cv = np.mean(cvscores)

IndexError: tuple index out of range

g)
--

In [16]:
n_decay = 10
lear_decay = np.logspace(-6,0,n_decay)
sgd = SGD(lr=0.2, decay=1e-6)

In [17]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))


/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: 3990076786485350814898454528.0000 - val_loss: 30777107681161577890137006669824.0000
Epoch 2/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan                       
Epoch 3/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 4/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 5/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 6/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 7/300
404/404 [==============================

In [18]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))


/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 2/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 3/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 4/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 5/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 6/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 7/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan

h)
--

In [33]:
n_decay = 21
momentum = np.linspace(0,1,n_decay)
sgd=SGD(lr=0.2,momentum=0.9)

In [34]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))



/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: 7973174141161343127476240384.0000 - val_loss: 24273109196023601857574920519680.0000
Epoch 2/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan                       
Epoch 3/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 4/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 5/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 6/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan

El resultado final que se observa es similar al observado en el punto c, viendo una convergencia de forma constante pero lenta.

In [35]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))


/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Train on 404 samples, validate on 102 samples
Epoch 1/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 2/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 3/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 4/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 5/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 6/300
404/404 [==============================] - 0s - loss: nan - val_loss: nan
Epoch 7/300
404/404 [==============================] - ETA: 0s - loss: nan

i)
--

In [36]:
n_batches = 21
batch_sizes = np.round(np.linspace(1,X_train_scaled.shape[0],n_batches))

In [37]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

model.fit(X_train_scaled.as_matrix(),y_train_scaled.as_matrix(),batch_size=50,nb_epoch=300)


/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(200, kernel_initializer="uniform", input_dim=14)`
  from ipykernel import kernelapp as app
/Users/ricardocarrasco/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(1, kernel_initializer="uniform")`


Epoch 1/300
404/404 [==============================] - 0s - loss: 44978.9883   
Epoch 2/300
404/404 [==============================] - 0s - loss: 11025.6185     
Epoch 3/300
404/404 [==============================] - 0s - loss: 4812.5682      
Epoch 4/300
404/404 [==============================] - 0s - loss: 2716.5407     
Epoch 5/300
404/404 [==============================] - 0s - loss: 1510.9406    
Epoch 6/300
404/404 [==============================] - 0s - loss: 1023.5479    
Epoch 7/300
404/404 [==============================] - 0s - loss: 934.8619      

In [ ]:
model = Sequential()
model.add(Dense(200, input_dim=X_train_scaled.shape[1], init='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, init='uniform'))
model.add(Activation('linear'))

model.compile(optimizer=sgd, loss='mean_squared_error')

hist = model.fit(X_train_scaled.as_matrix(), y_train_scaled.as_matrix(), nb_epoch=300, verbose=1, validation_data=(X_test_scaled.as_matrix(), y_test_scaled.as_matrix()))

